File: test/merge.test.js
// node --test  — pins the upsert merge a subdomain rescan does before writing
// CTEMData. Pure function, no Mongo, no scanners.
const { test } = require('node:test');
const assert = require('node:assert/strict');
const { mergeSubdomainResults, hostKey, belongsToDomain } = require('../index.js');

const by = (list) => Object.fromEntries(list.map((e) => [e.sub, e]));

test('hostKey keeps www., belongsToDomain ignores it', () => {
  assert.equal(hostKey('WWW.Acme.Test.'), 'www.acme.test');
  assert.equal(belongsToDomain('www.acme.test', 'acme.test'), true);
  assert.equal(belongsToDomain('acme.test', 'www.acme.test'), true);
  assert.equal(belongsToDomain('acme.testing', 'acme.test'), false);
});

test('rescan upserts: updates found entries, keeps unfound ones, creates new ones, leaves other domains alone', () => {
  const existing = [
    { _id: 'id-www', sub: 'www.acme.test', ip: '10.0.0.1', status: 'Active', assetCriticality: 'High', ownerEmail: 'o@acme.test', source: 'scan', sslGrade: 'A', sslDaysRemaining: 40 },
    { _id: 'id-vpn', sub: 'vpn.acme.test', ip: '', status: 'Pending', source: 'client-request', addedAt: new Date('2026-09-01'), addedBy: 'admin@x' },
    { _id: 'id-old', sub: 'old.acme.test', ip: '10.0.0.9', status: 'Active', source: 'scan' },
    { _id: 'id-shop', sub: 'shop.acme.co', ip: '10.5.5.5', status: 'Active', source: 'scan' },
    { _id: 'id-un', sub: 'lonely.other.net', ip: '', status: 'Pending', source: 'manual', rootDomain: '' },
  ];
  const scanned = [
    { sub: 'www.acme.test', ip: '10.9.9.9', status: 'Active', assetCriticality: 'Low', sslGrade: 'N/A', sslDaysRemaining: null },
    { sub: 'new.acme.test', ip: '10.1.1.1', status: 'Active', assetCriticality: 'Low', sslGrade: 'N/A', sslDaysRemaining: null },
    { sub: 'noip.acme.test', ip: '', status: 'Active', assetCriticality: 'Low', sslGrade: 'N/A', sslDaysRemaining: null },
    { sub: 'new.acme.test', ip: '10.1.1.1' }, // scanner repeated a name
  ];
  const r = mergeSubdomainResults(existing, scanned, 'acme.test');
  assert.deepEqual({ kept: r.kept, updated: r.updated, created: r.created, retained: r.retained }, { kept: 2, updated: 1, created: 2, retained: 2 });
  const m = by(r.merged);
  assert.equal(r.merged.length, 7);

  // found again → fresh ip, admin-owned + SSL + _id preserved, criticality NOT reset to Low
  assert.equal(m['www.acme.test'].ip, '10.9.9.9');
  assert.equal(m['www.acme.test']._id, 'id-www');
  assert.equal(m['www.acme.test'].assetCriticality, 'High');
  assert.equal(m['www.acme.test'].ownerEmail, 'o@acme.test');
  assert.equal(m['www.acme.test'].sslGrade, 'A');
  assert.equal(m['www.acme.test'].sslDaysRemaining, 40);
  assert.equal(m['www.acme.test'].rootDomain, 'acme.test');

  // not found this run → kept exactly as it was (the client-requested one and the stale scan one)
  assert.deepEqual(m['vpn.acme.test'], existing[1]);
  assert.deepEqual(m['old.acme.test'], existing[2]);

  // new → created as a scan entry
  assert.equal(m['new.acme.test'].source, 'scan');
  assert.ok(m['new.acme.test'].addedAt instanceof Date);
  assert.equal(m['new.acme.test'].rootDomain, 'acme.test');

  // no IP → Inactive, not blindly Active
  assert.equal(m['noip.acme.test'].status, 'Inactive');

  // other root domain and Unassigned entries untouched
  assert.deepEqual(m['shop.acme.co'], existing[3]);
  assert.deepEqual(m['lonely.other.net'], existing[4]);
});

test('www.acme.test and acme.test are distinct entries', () => {
  const existing = [
    { sub: 'acme.test', ip: '1.1.1.1', status: 'Active', assetCriticality: 'Critical', source: 'scan' },
    { sub: 'www.acme.test', ip: '1.1.1.2', status: 'Active', assetCriticality: 'Low', source: 'scan' },
  ];
  const scanned = [{ sub: 'acme.test', ip: '1.1.1.1' }, { sub: 'www.acme.test', ip: '1.1.1.2' }];
  const m = by(mergeSubdomainResults(existing, scanned, 'acme.test').merged);
  assert.equal(m['acme.test'].assetCriticality, 'Critical');
  assert.equal(m['www.acme.test'].assetCriticality, 'Low');
});

test('an Unassigned entry that a later root-domain scan finds is attributed to that domain', () => {
  const existing = [{ sub: 'api.acme.test', ip: '', status: 'Pending', source: 'client-request', rootDomain: '' }];
  const r = mergeSubdomainResults(existing, [{ sub: 'api.acme.test', ip: '2.2.2.2' }], 'acme.test');
  assert.equal(r.updated, 1);
  assert.equal(r.merged[0].rootDomain, 'acme.test');
  assert.equal(r.merged[0].source, 'client-request');
  assert.equal(r.merged[0].status, 'Active');
});

test('first scan on an empty tenant just creates', () => {
  const r = mergeSubdomainResults(undefined, [{ sub: 'a.acme.test', ip: '1.1.1.1' }], 'acme.test');
  assert.equal(r.created, 1);
  assert.equal(r.merged.length, 1);
});

File: internal/modules/subdomain/merge_test.go
package subdomain

import (
	"testing"
	"time"
)

func TestMerge_UpsertsWithoutLosingAdminOwnedData(t *testing.T) {
	now := time.Now()
	addedAt := now.Add(-24 * time.Hour)

	existing := []Subdomain{
		{Sub: "www.acme.test", IP: "10.0.0.1", Status: "Active", AssetCriticality: "High", OwnerEmail: "o@acme.test", Source: "scan"},
		{Sub: "vpn.acme.test", IP: "", Status: "Pending", Source: "client-request", AddedAt: &addedAt, AddedBy: "admin@x"},
		{Sub: "old.acme.test", IP: "10.0.0.9", Status: "Active", Source: "scan"},
		{Sub: "shop.acme.co", IP: "10.5.5.5", Status: "Active", Source: "scan"},
	}
	fresh := []Subdomain{
		{Sub: "www.acme.test", IP: "10.9.9.9"},
		{Sub: "new.acme.test", IP: "10.1.1.1"},
		{Sub: "noip.acme.test", IP: ""},
		{Sub: "new.acme.test", IP: "10.1.1.1"},
	}

	got := Merge(existing, fresh, "acme.test", now)

	if got.Created != 2 || got.Updated != 1 || got.Retained != 2 || got.Kept != 1 {
		t.Fatalf("counts: %+v", got)
	}
	if len(got.Merged) != 6 {
		t.Fatalf("merged %d entries, want 6", len(got.Merged))
	}

	byHost := map[string]Subdomain{}
	for _, s := range got.Merged {
		byHost[s.Sub] = s
	}

	if w := byHost["www.acme.test"]; w.IP != "10.9.9.9" || w.AssetCriticality != "High" || w.OwnerEmail != "o@acme.test" || w.RootDomain != "acme.test" {
		t.Errorf("www.acme.test: %+v", w)
	}
	if v := byHost["vpn.acme.test"]; v.Source != "client-request" || v.Status != "Pending" || v.IP != "" || v.AddedBy != "admin@x" {
		t.Errorf("vpn.acme.test should be retained untouched: %+v", v)
	}
	if _, ok := byHost["old.acme.test"]; !ok {
		t.Error("old.acme.test should have been retained")
	}
	if s := byHost["shop.acme.co"]; s.IP != "10.5.5.5" {
		t.Errorf("shop.acme.co must be untouched: %+v", s)
	}
	if n := byHost["new.acme.test"]; n.Source != "scan" || n.AddedAt == nil {
		t.Errorf("new.acme.test: %+v", n)
	}
	if byHost["noip.acme.test"].Status != "Inactive" {
		t.Errorf("noip.acme.test status = %q, want Inactive", byHost["noip.acme.test"].Status)
	}
}

func TestMerge_ApexAndWWWAreDistinctEntries(t *testing.T) {
	existing := []Subdomain{
		{Sub: "acme.test", AssetCriticality: "Critical", Source: "scan"},
		{Sub: "www.acme.test", AssetCriticality: "Low", Source: "scan"},
	}
	fresh := []Subdomain{{Sub: "acme.test", IP: "1.1.1.1"}, {Sub: "www.acme.test", IP: "1.1.1.2"}}

	byHost := map[string]Subdomain{}
	for _, s := range Merge(existing, fresh, "acme.test", time.Now()).Merged {
		byHost[s.Sub] = s
	}
	if byHost["acme.test"].AssetCriticality != "Critical" || byHost["www.acme.test"].AssetCriticality != "Low" {
		t.Fatalf("apex and www must not collide: %+v", byHost)
	}
}

func TestHostKeyAndDomainAttribution(t *testing.T) {
	if got := hostKey("WWW.Acme.Test."); got != "www.acme.test" {
		t.Errorf("hostKey = %q, want www.acme.test", got)
	}
	if !belongsToDomain("www.acme.test", "acme.test") {
		t.Error("www.acme.test belongs to acme.test")
	}
	if !belongsToDomain("acme.test", "www.acme.test") {
		t.Error("attribution ignores a leading www. on either side")
	}
	if belongsToDomain("acme.testing", "acme.test") {
		t.Error("acme.testing must not be attributed to acme.test")
	}
}

func TestMerge_FirstScanOnEmptyTenant(t *testing.T) {
	got := Merge(nil, []Subdomain{{Sub: "a.acme.test", IP: "1.1.1.1"}}, "acme.test", time.Now())
	if got.Created != 1 || len(got.Merged) != 1 {
		t.Fatalf("%+v", got)
	}
}
